### PR TITLE
Refactor ObjectManager wait logic to WaitManager

### LINKED ISSUE #21385: [Core] [Bug] non_streaming_shuffle_1tb_1000_partition test hangs when lease parallelism is 10
### Search before asking

- [X] I searched the [issues](https://github.com/ray-project/ray/issues) and found no similar issues.

### Ray Component

Ray Core

### What happened + What you expected to happen

The issue is hanging actor method with large object arguments `actor.method.remote(large_object)` : actor core worker pulls the arguments by calling `NodeManager.ProcessWaitForDirectActorCallArgsRequestMessage` which basically does two things `AsyncResolveObjects` and `object_manager_.Wait` . `ObjectManager::Wait` basically calls `SubscribeRemainingWaitObjects` which calls `object_directory_->SubscribeObjectLocations`, the wait is complete only when the `local_objects_` has the waiting objects when the subscription notification is fired:
```
RAY_CHECK_OK(object_directory_->SubscribeObjectLocations(
          wait_id, object_id, wait_state.owner_addresses[object_id],
          [this, wait_id](const ObjectID &subscribe_object_id, ...) {
            ...
            if (local_objects_.count(subscribe_object_id) > 0) {
              wait_state.remaining.erase(subscribe_object_id);
              wait_state.found.insert(subscribe_object_id);
              wait_state.requested_objects.erase(subscribe_object_id);
              RAY_CHECK_OK(object_directory_->UnsubscribeObjectLocations(
                  wait_id, subscribe_object_id));
              if (wait_state.found.size() >= wait_state.num_required_objects) {
                WaitComplete(wait_id);
              }
            }
          }));
```
In the hanging case, when the notification is called, the waiting object is already deleted from `local_objects_` so `WaitComplete` is never called.
The waiting object is deleted because after it’s get pulled from remote, `DependencyManager::HandleObjectLocal` is called and the active pulling request is considered done and is removed through `PullManager::DeactivatePullBundleRequest `which unpin the object and then it’s deleted due to memory pressure.

So the sequence of events is (code is https://github.com/ray-project/ray/pull/20404/files)
```
[2022-01-03 14:05:57,402 I 111 111] node_manager.cc:1461: jjyao object_manager wait [ffffffffffffffffffffffffffffffffffffffff0200000001000000, ffffffffffffffffffffffffffffffffffffffff0200000002000000]
21232[2022-01-03 14:05:57,402 I 111 111] object_manager.cc:568: jjyao LookupLocations ffffffffffffffffffffffffffffffffffffffff0200000001000000
21483[2022-01-03 14:05:57,447 I 111 111] object_manager.cc:585: jjyao ffffffffffffffffffffffffffffffffffffffff0200000001000000 is not local
21484[2022-01-03 14:05:57,447 I 111 111] object_manager.cc:587: jjyao Wait request 81826c8c36b0b9a283aa496a34567c84f2beddbd7201014abba661b0: 1 locations found for object ffffffffffffffffffffffffffffffffffffffff0200000001000000
21486[2022-01-03 14:05:57,447 I 111 111] object_manager.cc:613: jjyao Wait request 81826c8c36b0b9a283aa496a34567c84f2beddbd7201014abba661b0: subscribing to object ffffffffffffffffffffffffffffffffffffffff0200000001000000
21618[2022-01-03 14:05:57,469 I 111 111] object_manager.cc:623: jjyao SubscribeObjectLocations returned ffffffffffffffffffffffffffffffffffffffff0200000001000000
21619[2022-01-03 14:05:57,469 I 111 111] pull_manager.cc:172: jjyao Activating pull for object ffffffffffffffffffffffffffffffffffffffff0200000001000000 1
21620[2022-01-03 14:05:57,469 I 111 111] object_manager.cc:243: jjyao SendPullRequest ffffffffffffffffffffffffffffffffffffffff0200000001000000
21781[2022-01-03 14:05:57,939 I 111 111] object_manager.cc:167: jjyao Object added ffffffffffffffffffffffffffffffffffffffff0200000001000000
21782[2022-01-03 14:05:57,939 I 111 111] pull_manager.cc:617: jjyao Pinned newly created object ffffffffffffffffffffffffffffffffffffffff0200000001000000
21783[2022-01-03 14:05:57,939 I 111 111] dependency_manager.cc:294: jjyao Canceling pull for wait request of object ffffffffffffffffffffffffffffffffffffffff0200000001000000 request: 1
21785[2022-01-03 14:05:57,939 I 111 111] pull_manager.cc:202: jjyao Deactivating pull for object ffffffffffffffffffffffffffffffffffffffff0200000001000000
21787[2022-01-03 14:05:57,939 I 111 111] dependency_manager.cc:41: jjyao Object ffffffffffffffffffffffffffffffffffffffff0200000001000000 no longer needed
21878[2022-01-03 14:05:57,953 I 111 111] object_manager.cc:196: jjyao Object deleted ffffffffffffffffffffffffffffffffffffffff0200000001000000
22553[2022-01-03 14:05:58,055 I 111 111] object_manager.cc:623: jjyao SubscribeObjectLocations returned ffffffffffffffffffffffffffffffffffffffff0200000001000000
74782[2022-01-03 14:07:36,195 I 111 111] object_manager.cc:623: jjyao SubscribeObjectLocations returned ffffffffffffffffffffffffffffffffffffffff0200000001000000
74783[2022-01-03 14:07:36,198 I 111 111] object_manager.cc:623: jjyao SubscribeObjectLocations returned ffffffffffffffffffffffffffffffffffffffff0200000001000000
```

### Versions / Dependencies

master

### Reproduction script

non_streaming_shuffle_1tb_1000_partition test with lease parallelism 10.

### Anything else

_No response_

### Are you willing to submit a PR?

- [X] Yes I am willing to submit a PR!

### PR DESCRIPTION
**Pull request by jjyao**
## Why are these changes needed?
- This PR moves the `ObjectManager::Wait` related logic to a separate WaitManager class.
- Fix the wait hang issue by not  relying on the async object location notification, but checking if wait is complete when the local object is added, at that time the object is guaranteed to be local.

## Related issue number
Closes #21385

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Review by stephanie-wang — APPROVED** [review]
Nice, this looks a lot cleaner!

Could you clean up the PR description a bit to make it shorter?

**Comment by ericl** [comment]
+1 can you also link to the issue this is solving / impact?

**Comment by jjyao** [comment]
It's ready for review.

File: src/ray/raylet/wait_manager.h [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+
+#pragma once
+
+#include "ray/common/asio/instrumented_io_context.h"
+#include "ray/common/id.h"
+
+namespace ray {
+namespace raylet {
+
+/// This class is responsible for handling wait requests (e.g. `ray.wait`).
```
Comment (stephanie-wang) [inline src/ray/raylet/wait_manager.h:23]: Can you specify that it's only used for `fetch_local=true` requests?

File: src/ray/raylet/wait_manager.h [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+/// It is not thread safe and is expected to run ONLY in
+/// the NodeManager.io_service_ thread.
+class WaitManager {
+ public:
+  WaitManager(instrumented_io_context *io_service,
```
Comment (stephanie-wang) [inline src/ray/raylet/wait_manager.h:28]: It' nicer for testing to pass a lambda for creating a timer. You can look at PullManager for an example.

**Review by stephanie-wang — COMMENTED** [review]
Thanks for the cleanup and the unit tests! This is so much nicer.

Just a couple comments for polish before we merge.

**Review by stephanie-wang — APPROVED** [review]

**Pull request merged**